README.md

# ffmpeg-ps1

- [About powershell-ffmpeg](#about-ffmpeg-ps1)
  - [Requirements](#requirements)
- [User Case: Media Server](#user-case-media-server)
- [Simplified Script Flow](#simplified-script-flow)
- [Config In Detail](#configuration)

## About ffmpeg-ps1

This repo contains a script that i've put together using powersell as the base language. The main purpose of this script is to use [FFmpeg](https://ffmpeg.org/) to encode/convert/compress video files in media folders. This allows you to standardize file sizes throughout your media library, also enforcing maximum resolutions for longer term storage or consistency.

### Requirements

- FFMPEG needs to be installed and configured to work with CLI in powershell.

  - [How to install ffmpeg](https://gist.github.com/barbietunnie/47a3de3de3274956617ce092a3bc03a1) (Github)

- Powershell ([How to install](https://learn.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-7.3)) - it is availble in multiple operating systems including but not limited to:
  - Windows
  - Windows Server
  - macOS
  - Debian

## User Case: Media Server

Our user has a home media server that runs on Windows PC in which all users within their household can stream and watch any of their stored anime, movies and TV shows. Over time they noticed that they are experiencing a couple of key issues.

- When saving the media to their server newer files are coming in increasingly larger file sizes and resolutions
- They are experiencing increased load trying to transocde media on the fly with their media client that is not a direct stream. Examples include:
  - Downscaling
  - Subtitles that are being burned in
  
 While looking for solutions they only want to install the bare minimum to get a working solution. Using [PowerShell](https://learn.microsoft.com/en-us/powershell/scripting/overview?view=powershell-7.3) they are able to create and run a shell intigrated script that uses [FFmpeg](https://ffmpeg.org/about.html) to convert their oversized video files to a file format and size that fits their needs. This script is run through their operating systems task automation service ([Task Scheduler](https://learn.microsoft.com/en-us/windows/win32/taskschd/task-scheduler-start-page)/[Cron Job](https://en.wikipedia.org/wiki/Cron)/other) they can have the script run atumatically on a regular basis to maintain their library.

## Simplified Script Flow

1. Powershell begins scanning all indicated directories. Each file and folder under that path is added to contents.txt
1. Powershell will begin scanning through each file individually, skipping folders, and and attempts to determine current bitrate, resolution and calculate if the file requires encoding when compared to presets.
1. Once file data has been determined it will begin encoding the new file.
1. When the file encode is complete, it will dispose of the source file and move the new file to the same directory with the same naming convention.

## Configuration

When it come to configuring for use you may want to modify certain options depending on your intented use. For example, I primarily use this script in the background through a scheduled task. As this is run automatically I do not need any GUI based items enabled so those would be set to false. A copy of the default configuration can be seen below.

```powershell
# Initial Config
  Set-PSDebug -Off # Disables debugging, only shows verbose (if enabled) and running commands
  $bVerbose = $False # If `$True` verbose messages are enabled in the console while script is running.
  $bDisableStatus = $True # Set to true if you wish to disable the calculating and displaying of status/progress bars in the script (can increase performance)
  $sRootPath = "D:" # This is the root file path you want power-shell to begin scanning for media if you are wanting to scan all child items of this directory. *This becomes very important if you have `$bRecursiveSearch` set to `$False`*.
  $sEncodePath = "D:\Encode\" # The folder/path where you wish to remporarely store encodes while they are being processed. *It is recommended to use a different location from any other files.*
  $sExportedDataPath = $sScriptPath # The folder/path where you want the exported files to be generated. 'Exported files' does not include encodes.
  $bRecursiveSearch = $False # This controls if you wish to scan the entire root folder specified in `$sRootPath` for content. If `$True`, all files, folders and subfolders will be subject to at least a scan attempt. If `$False`, only the folders indicated in `$sDirectoriesCSV` will be subject to a recursive scan.
      $sDirectoriesCSV = "D:\Anime\,D:\TV\,D:\Movies\" # If you want to only have power-shell scan specific folders for media, you can indicate all paths in this variable using CSV style formatting.
  # Script Testing
  $bTestFlow = $True # If `$True` the script will follow through all normal steps without actually encoding, moving, or deleting video files
  $bTestSingleFile = $False # If `$True` Enables test mode. Test mode only scans and encodes a single source path defined in `$sTestSinglePath`. Destination file is saved to your `$sExportedDataPath`.
  $sTestSinglePath = "D:\Downloads\TestFile.mkv" # Source Path to file you want to test the script on.
# Execution Limits
  $iLimitQueue = 0 #No limit = `0`. Limits the number of files that are encoded per execution. Once this number has been reached it will stop. It can be stopped early if also used in conjunction with `$iEncodeHours`.
  $iEncodeHours = 0 #No limit = `0`. Limits time in hours in you allow a single script execution to run. End time will be obtained before scanning starts. It will then check that the time has not been exceeded before each encode begins.
  If ($iLimitQueue -ne 0 -or $iEncodeHours -ne 0){$bEncodeLimit = $True}Else{$bEncodeLimit = $False} # If either of the limit controllers contain values above 0, then this is marked as `$True`
# Exported Data
  $bEncodeOnly = $True # When this is `$True`, only items identified as "needing encode" as per the `Detect Medtadata > Video Metadata > Check if encoding needed` section. If `$False` then all items will be added to the CSV regardless if encoding will take place for the file or not. *This does not change whether or not the file **will** be encoded, only if it is logged in the generated CSV file*
  $bDeleteCSV = $False # If `$False` then `contents.csv` will be deleted after the script is finished. If `$True` then `contents.csv` will **not** be deleted after the script is finished. Instead the next time it runs it will be written over.
  $bAppendLog = $True # If `$False` then when a new encoding session begins, the contents of `encode.log` are cleared. If `$True` then the contents of said text file will append until cleared manually.
  $bDeleteContents = $True # If `$False` then the `contents.txt` file generated at scanning will not be deleted after `contents.csv` is created. If `$True` then `contents.txt` will be deleted after `contents.csv` is created.
  $bDeleteSource = $True # If `$True` then the source video file for each encode will be deleted entirely after encode is complete. If `$False` then the source video file is moved to `$sEncodePath\old`
# Encode Settings
  $bRemoveBeforeScan = $True # If `$True` then  all files in `$sEncodePath` are deleted prior to initiated a scan for media
  $iEncodeOrder = 3
      # If `0` then script will stop running and not continue past it's current step
      # If `1` then script will log data to CSV files but not encode any files
      # If `2` then script will encode files after CSV is generated
      # If `3` then script will encode files as they are detected without adding them to a CSV file
# Video Reolution Limits
  $iResMin = 480 # Lower limit of pixel height detection
  $iResMax = 1000 # Upper limit of pixel height detection
  $sResMin = "640x480" # If pixel height is less than `$iResMin` then encode at this quality
  $sResMed = "1280x720" # If pixel height is greater than `$iResMin` but less than `$iResMax` encode at this quality
  $sResMax = "1920x1080" # If pixel height is greater than `$iResMax` encode at this quality
# Bitrate Limit
  $iBitRateSD = 1000 # bitrate in kbps for video files with a vertical pixel count < 480
  $iBitRate720 = 2000 # bitrate in kbps for video files with a verticle pixel count > 480 and a pixel count < 1000
  $iBitRate1080 = 2500 # bitrate in kbps for video files with a verticle pixel count > 1000
# ffmpeg flags in order of use
  # `-i <inputpath>` input path for source file 
  # `-b <int>` video bitrate. Source: $_.T_Bits_Ps
  # `-maxrate <int>` maximum bitrate tolerance (in bits/s). Requires bufsize to be set. (from INT_MIN to INT_MAX) (default 0). Source: $_.T_Bits_Ps
  # `-minrate <int>` minimum bitrate tolerance (in bits/s). Most useful in setting up a CBR encode. It is of little use otherwise. (from INT_MIN to INT_MAX) (default 0). Source: $_.T_Bits_Ps
  $sff_ab = '64k'# `-ab <str>` bitrate (in bits/s) (from 0 to INT_MAX) (default 128000). Source: User defined
  $sff_vcodec = 'libx264' # `-vcodec <str>` force video codec (‘copy’ to copy stream). Source: User defined
  $sff_acodec = 'aac' # `-acodec <str>` force audio codec (‘copy’ to copy stream). Source: User defined
  $sff_strict = 2 # `-strict <int>` ED.VA… how strictly to follow the standards (from INT_MIN to INT_MAX) (default 0). Source: User defined
  $sff_ac = 2 # `-ac <int>` channels set number of audio channels. Source: User defined
  $sff_ar = 44100 # `-ar <int>` rate set audio sampling rate (in Hz). Source: User defined
  # `-s <str>` size set frame size (WxH or abbreviation). Source: $_.T_height 
  $sff_map = 0 # `-map <int>` -map [-]input_file_id[:stream_specifier][,sync_file_id[:stream_s set input stream mapping. Source: User defined
  $sff_threads = 1 # `-threads <int>` (from 0 to INT_MAX) (default 1). Source: User Defined
  # `-v <string>` set logging level. Source: Built into command
  # `-stats` print progress report during encoding
  # `<outputpath>` output path of file being created. Source: $n_path
```


changelog.md
# changelog

## Pending Changes

- Added config option that allows you to sort the exported `contents.csv` file by specific columns

## v1.3

- changelog does not exist


Encode.ps1
$sScriptPath = split-path -parent $MyInvocation.MyCommand.Definition # Gets the path of the script file being executed
try{ . ./config.ps1}
catch{ . .\config.ps1}
#Functions
    Function EncodeLog($string){
        $sTimeStamp = Get-Date -Format "yyyy-MM-dd HH:mm"
        $sLogStamp = "$sTimeStamp $string"
        write-output $sLogStamp | add-content $sExportedDataPath\encode.log
        Write-Verbose -Message $string
    }
    Function ErrorLog{
        $sTimeStamp = Get-Date -Format "yyyy-MM-dd HH:mm"
        $sLogStamp = "$sTimeStamp | Error $sErrorTask | $sErrorMessage"
        write-output $sLogStamp | add-content $sExportedDataPath\error.log
    }
    Function EncodeCSV {
        # Adds current scanned item to Encode.csv if it meets the requirements
        [pscustomobject]@{
            Bits_Ps = $iBits
            height = $iHeight
            T_Bits_Ps = $iScaleBits
            T_height = $theight
            Encode = $bEncode
            Path = $sContentsLine
        }
    }
    # Begins the encoding process of all items marked "Encode = TRUE" in contents.csv
    Function BeginEncode {
        #Begin encoding
        If ($bAppendLog -eq $False) {Clear-Content -Path $sExportedDataPath\encode.log} # Clears log file at start of encode if true, otherwise appends continuously
        If ($bDisableStatus -eq $False) {
            $iSteps = (get-content $sExportedDataPath\contents.csv).length
        } # If bDisableStatus is False then updates the gui terminal with status bar
        $Istep = 0
        #Loop through contents.csv and encode each file identified
            Import-Csv $sExportedDataPath\contents.csv | ForEach-Object {
                if($bEncodeAfterScan -eq $True){
                    try{
                        if ($($_.encode) -eq "TRUE") {
                            If ($bDisableStatus -eq $False) {
                                $iPercent = ($iStep/$iSteps)*100
                            } # If bDisableStatus is False then updates the gui terminal with status bar
                            #Collect file details
                                $sFilename = Get-ChildItem $($_.path)
                                $sBasename = $sFilename.BaseName #to get name only
                                If ($bTest -eq $True) {$outputpath = $sExportedDataPath+$sBasename+".mkv"} Else {$outputpath = $sEncodePath+$sBasename+".mkv"}
                                
                                $sInputContainer = split-path -path $($_.path)
                                If ($bDisableStatus -eq $False) {Write-Progress -Activity "Encoding: $iStep/$iSteps" -Status "$sFilename" -PercentComplete $iPercent} # If bDisableStatus is False then updates the gui terminal with status bar
                                Write-Verbose -Message "Working $sFilename"
                            #Create new encode
                                ffmpeg -i "$($_.path)" -b $($_.T_Bits_Ps) -maxrate $($_.T_Bits_Ps) -minrate $($_.T_Bits_Ps) -ab $sff_ab -vcodec $sff_vcodec -acodec $sff_acodec -strict $sff_strict -ac $sff_ac -ar $sff_ar -s $($_.T_height) -map $sff_map -y -threads $sff_threads -v quiet -stats $outputpath
                                #ffmpeg -i "$($_.path)" -b $($_.T_Bits_Ps) -maxrate $($_.T_Bits_Ps) -minrate $($_.T_Bits_Ps) -ab 64k -vcodec libx264 -acodec aac -strict 2 -ac 2 -ar 44100 -s $($_.T_height) -map 0 -y -threads 2 -v quiet -stats $outputpath
                            #Check thar files still exist before removal
                                $sSourcePath = Test-Path $($_.path)
                                $sDestPath = Test-Path $outputpath
                                $iDestSize = (Get-Item $outputpath).Length/1MB
                                if ($sDestPath -eq $True -and $sSourcePath -eq $True -and $iDestSize -gt 1) {
                                    #Remove input file
                                        If ($bDeleteSource -eq $True){
                                            # Delete input file if set to true
                                            remove-item $($_.path)
                                            EncodeLog("Deleted source file")
                                        }
                                        Else{
                                            # Otherwise relocate it
                                            Move-Item -Path $($_.path) -Destination "$sEncodePath\old" -Force
                                            EncodeLog("Moved source file to $sEncodePath\old")
                                        }
                                        
                                    #Move new file to original folder
                                        move-item $outputpath -Destination $sInputContainer
                                    #Populate log of encoded files
                                        $iTargetBits = ($($_.T_Bits_Ps)/1000)
                                        $iOriginBits = ($($_.Bits_Ps)/1000)
                                        EncodeLog "($iStep) $sBasename encoded in $($_.T_height)p at $iTargetBits kbp/s | Originally $iOriginBits kbp/s"
                                        Write-Verbose -Message "Complete"
                                        $iStep++
                                }
                                Else{
                                    If($iDestSize -lt 1){EncodeLog("Aborting file overwrite as encode file is less than 1MB")}
                                    Elseif($sDestPath -eq $False){EncodeLog("Aborting file overwrite as destination path does not exist anymore")}
                                    Else{EncodeLog("Aborting file overwrite as encode file does not exist anymore")}
                                }
                        }
                        If ($bDisableStatus -eq $False) {Write-Progress -Activity "Encoding: $iStep/$iSteps" -Status "$sFilename" -Completed} # If bDisableStatus is False then updates the gui terminal with status bar
                        If ($bEncodeLimit -eq $True){
                            # Checks current stats against encode limits
                            If ($iLimitQueue -gt 0 -and $iStep -ge $iLimitQueue){
                                # Confirms if limit of concurrent encodes has been reached
                                $bEncodeAfterScan = $False
                                EncodeLog("Encode Limit Reached as queue has processed $iLimitQueue file(s)")
                            }
                            If ($iEncodeHours -gt 0)
                            {
                                # Confirms the script is allowed to continue running at this time
                                $DateTime = Get-Date
                                If ($DateTime -ge $EndTime){
                                    $bEncodeAfterScan = $False
                                    EncodeLog("Encode Limit Reached as queue been reached the $iEncodeHours hour limit")
                                }
                            }
                        }
                    }
                    catch{
                        $sErrorTask = "Encoding $sBasename"
                        $sErrorMessage = $_
                        ErrorLog
                    }
                }
            }
    }
# End Fnctions
#Initialize
try {
    EncodeLog("Initializing")
    $sErrorTask = "Initializing"
    Set-Location $sRootPath # set directory of root folder for monitored videos
    $EndTime = Get-Date
    $EndTime = $EndTime.AddHours($iEncodeHours)
    If ($bVerbose -eq $True) {$VerbosePreference = "Continue"} Else {$VerbosePreference = "SilentlyContinue"} # If verbose is on, shows verbose messages in console
    If ($bRemoveBeforeScan -eq $True) {Remove-Item $sEncodePath -Include *.* -Recurse} # Remove old encodes
}
catch{
    $sErrorMessage = $_
    ErrorLog
} 
# Check folders before scanning
    try{
        EncodeLog("Checking Folders")
        $sErrorTask = "Checking Folders"
        If ((Test-Path -Path $sRootPath -PathType Container) -eq $False) {
            $sErrorMessage = "Root Path not found, aborting script"
            Write-Verbose -Message $sErrorMessage
            ErrorLog
            Exit
        }Else{Write-Verbose -Message "Root path found"}
        If ((Test-Path -Path $sEncodePath -PathType Container) -eq $False) {
            Write-Verbose -Message "Encode Path not found, creating folder"
            New-Item -ItemType "directory" -Path $sEncodePath
            #Test path again
            If ((Test-Path -Path $sEncodePath -PathType Container) -eq $False) {
                Write-Verbose -Message "Failed to create folder, redirecting to root path"
                $sEncodePath = $sRootPath
            }
        }Else{Write-Verbose -Message "Encode path found"}
    }
    catch{
        $sErrorMessage = $_
        ErrorLog
    }
# Start Scanning
    #Generate Contents
        #Generate Contents Lists and repeat based on number of directories
            try{
                EncodeLog("Scanning Files")
                $sErrorTask = "Generating Contents List"
                out-file $sExportedDataPath\contents.txt #create empty contents file
            If ($bTest -eq $True){
                $sTestPath | Add-Content $sExportedDataPath\contents.txt # If testmode active, export single path to contents.txt 
                #Otherwise follow default scan export
            }ElseIf ($bRecursiveSearch -eq $False){
                $sDirectoriesCSV.Split(",") | ForEach-Object {
                    Get-ChildItem -Path $_ -Recurse -Include "*" | ForEach-Object {$_.FullName} | Write-Output | Add-Content $sExportedDataPath\contents.txt
                }
            }Else{Get-ChildItem -Path $sRootPath -Recurse -Include "*" | ForEach-Object {$_.FullName} | Write-Output | Add-Content $sExportedDataPath\contents.txt}
            }
            catch{
                $sErrorMessage = $_
                ErrorLog
            }
    #Detect Metadata
        #Begin scanning files
            If ($bDisableStatus -eq $False) {$activity = "Collecting Metadata from files"} # If bDisableStatus is False then updates the gui terminal with status bar
        #Start grabbing metadata based on contents
            $iSteps = (get-content $sExportedDataPath\contents.txt).length
            $iStep = 0
            $iPercent = 0
            $ffmpeg =@(
                foreach($sContentsLine in Get-Content $sExportedDataPath\contents.txt){
                    try{
                        $sErrorTask = "Obtaining Metadata"
                        If ($bDisableStatus -eq $False) {Write-Progress -Activity $activity -Status "Progress:" -PercentComplete $iPercent} # If bDisableStatus is False then updates the gui terminal with status bar
                        #Check file folder and parent folder for ".skip" file to skip the encoding of these folders
                            $sFilePath = Split-Path -Path $sContentsLine
                            $sSkipPath = $sFilePath + "\.skip"
                            $sParentPath = Split-path -Parent $sContentsLine
                            $sParentSkipPath = $sParentPath + "\.skip"
                        #If skip file not found in either path then get video metadata
                            $bScanFile = $True # Reset ScanFile for each item in contents.txt

                            If ($bTest -eq $False) {
                                If ((Test-Path -Path $sSkipPath) -and (Test-Path -Path $sParentSkipPath)) {$bScanFile = $False}
                            } # Runs if test mode is off - Looks for a .skip file in either the source directory or parent directy. If skip file is found, do not attempt to scan/encode file
                            If (Test-Path -Path $sContentsLine -PathType Container) {$bScanFile = $False} # If path is to folder, do not attempt to scan/encode path
                            If ($bScanFile -eq $True) {
                        #Video Metadata
                            #$iBits = ffprobe "$sContentsLine" -v error -select_streams v:0  -show_entries stream_tags=BPS -of default=noprint_wrappers=1:nokey=1 #get the video kbps via tag (very accurate)
                            $iBits = ffprobe "$sContentsLine" -v quiet -show_entries format=bit_rate -of default=noprint_wrappers=1:nokey=1 #if tag blank then get via format (less accurate)
                            $iHeight = ffprobe "$sContentsLine"  -v quiet -select_streams v:0  -show_entries stream=height -of default=noprint_wrappers=1:nokey=1 # get video width
                        
                            # Logic for desired bitrate based on video height
                                if ([int]$iHeight -le 480) {
                                    $kbps = $iBitRate480
                                    $theight = "640x480"
                                }elseif ([int]$iHeight -ge 1000) {
                                    $kbps = $iBitRate1080
                                    $theight = "1920x1080"
                                }else {
                                    $kbps = $iBitRate720
                                    $theight = "1280x720"
                                }
                        
                            # Check if encoding needed
                                $iScaleBits = [int]$kbps*1000
                                If($bTest -eq $True) {$bEncode = $True} ElseIf ([int]$iBits -gt $iScaleBits*1.3) {$bEncode = $True} else {
                                    $bEncode = $False
                                    Write-Verbose -Message "Encoding determined not needed for path - $sContentsLine"
                                } # Check if bitrate is greater than target kbp/s if so mark for encode
                        
                            # Add data to array
                                If ($bTest -eq $True) {
                                    EncodeCSV
                                    Write-Verbose -Message "Adding to CSV as bTest is True $sContentsLine"
                                } #Encode test path even if it doesnt need it
                                ElseIf ($bEncodeOnly -eq $True) {
                                    #If encode only is true, only import items needing encode into csv
                                    If ($bEncode -eq $True) {
                                        EncodeCSV
                                        Write-Verbose -Message "Adding to CSV as bEncode is True - $sContentsLine"
                                    }
                                }Else {
                                    #If encode only is false, import all items into csv
                                    EncodeCSV
                                    Write-Verbose -Message "Adding to CSV as bEncode is False - $sContentsLine"
                                }

                        }Else {
                            Write-Verbose -Message "Skip file exists, or path is folder. Skipping - $sContentsLine"
                        }
                        If ($bDisableStatus -eq $False) {
                            $iStep++
                            $iPercent = ($iStep/$iSteps)*100
                        } # If bDisableStatus is False then updates the gui terminal with status bar
                    }
                    catch{
                        $sErrorMessage = $_
                        ErrorLog
                    }
                }
            )
#Export CSV
    EncodeLog("Exporting File List")
    $ffmpeg | Export-Csv -Path $sExportedDataPath\contents.csv #export array to csv
    If ($bDisableStatus -eq $False) {Write-Progress -Activity $activity -Status "Ready" -Completed} # If bDisableStatus is False then updates the gui terminal with status bar         
    If ($bDeleteContents -eq $True) {remove-item $sExportedDataPath\contents.txt}
    If ($bEncodeAfterScan -eq $True) {
        EncodeLog("Begining Encode")
        BeginEncode
        EncodeLog("Encode Finished")
    } #Begin video encode if turned on in config
    If ($bDeleteCSV -eq $True) {remove-item $sExportedDataPath\contents.csv} #Remove contents csv if marked true in config

config.ps1
#config
    # Initial Config
    Set-PSDebug -Off # Disables debugging, only shows verbose (if enabled) and running commands
    $bVerbose = $True # If `$True` verbose messages are enabled in the console while script is running.
    $bTest = $False # If `$True` Enables test mode. Test mode only scans and encodes a single source path defined in `$sTestPath`. Destination file is saved to your `$sExportedDataPath`.
    $sTestPath = "D:\Downloads\TestFile.mkv" # Source Path to file you want to test the script on.
    $sRootPath = "D:" # This is the root file path you want power-shell to begin scanning for media if you are wanting to scan all child items of this directory. *This becomes very important if you have `$bRecursiveSearch` set to `$False`*.
    $sEncodePath = "D:\Encode" # The folder/path where you wish to remporarely store encodes while they are being processed. *It is recommended to use a different location from any other files.*
    $sExportedDataPath = $sScriptPath # The folder/path where you want the exported files to be generated. 'Exported files' does not include encodes.
    $bRecursiveSearch = $False # This controls if you wish to scan the entire root folder specified in `$sRootPath` for content. If `$True`, all files, folders and subfolders will be subject to at least a scan attempt. If `$False`, only the folders indicated in `$sDirectoriesCSV` will be subject to a recursive scan.
    $sDirectoriesCSV = "D:\Anime\,D:\TV\,D:\Movies\" # If you want to only have power-shell scan specific folders for media, you can indicate all paths in this variable using CSV style formatting.
    $bDisableStatus = $True # Set to true if you wish to disable the calculating and displaying of status/progress bars in the script (can increase performance)
# Limits
    $iLimitQueue = 0 #No limit = `0`. Limits the number of files that are encoded per execution. Once this number has been reached it will stop. It can be stopped early if also used in conjunction with `$iEncodeHours`.
    $iEncodeHours = 0 #No limit = `0`. Limits time in hours in you allow a single script execution to run. End time will be obtained before scanning starts. It will then check that the time has not been exceeded before each encode begins.
    If ($iLimitQueue -ne 0 -or $iEncodeHours -ne 0){$bEncodeLimit = $True}Else{$bEncodeLimit = $False} # If either of the limit controllers contain values above 0, then this is marked as `$True`
# Exported Data
    $bEncodeOnly = $True # When this is `$True`, only items identified as "needing encode" as per the `Detect Medtadata > Video Metadata > Check if encoding needed` section. If `$False` then all items will be added to the CSV regardless if encoding will take place for the file or not. *This does not change whether or not the file **will** be encoded, only if it is logged in the generated CSV file*
    $bDeleteCSV = $False # If `$False` then `contents.csv` will be deleted after the script is finished. If `$True` then `contents.csv` will **not** be deleted after the script is finished. Instead the next time it runs it will be written over.
    $bAppendLog = $True # If `$False` then when a new encoding session begins, the contents of `encode.log` are cleared. If `$True` then the contents of said text file will append until cleared manually.
    $bDeleteContents = $True # If `$False` then the `contents.txt` file generated at scanning will not be deleted after `contents.csv` is created. If `$True` then `contents.txt` will be deleted after `contents.csv` is created.
    $bDeleteSource = $True # If `$True` then the source video file for each encode will be deleted entirely after encode is complete. If `$False` then the source video file is moved to `$sEncodePath\old`
# Encode Confige
    $bRemoveBeforeScan = $True # If `$True` then  all files in `$sEncodePath` are deleted prior to initiated a scan for media
    $bEncodeAfterScan = $False # If `$False` then once the CSV is created the script skips the encoding process entirely. If `$True` then the script will encode all identified files after the CSV is generated.
# Bitrate Configuration
    $iBitRate480 = 1000 # bitrate in kbps for video files with a vertical pixel count < 480
    $iBitRate720 = 2000 # bitrate in kbps for video files with a verticle pixel count > 480 and a pixel count < 1000
    $iBitRate1080 = 2500 # bitrate in kbps for video files with a verticle pixel count > 1000
# ffmpeg flags in order of use
    # `-i <inputpath>` input path for source file 
    # `-b <int>` video bitrate. Source: $_.T_Bits_Ps
    # `-maxrate <int>` maximum bitrate tolerance (in bits/s). Requires bufsize to be set. (from INT_MIN to INT_MAX) (default 0). Source: $_.T_Bits_Ps
    # `-minrate <int>` minimum bitrate tolerance (in bits/s). Most useful in setting up a CBR encode. It is of little use otherwise. (from INT_MIN to INT_MAX) (default 0). Source: $_.T_Bits_Ps
    $sff_ab = '64k'# `-ab <str>` bitrate (in bits/s) (from 0 to INT_MAX) (default 128000). Source: User defined
    $sff_vcodec = 'libx264' # `-vcodec <str>` force video codec (‘copy’ to copy stream). Source: User defined
    $sff_acodec = 'aac' # `-acodec <str>` force audio codec (‘copy’ to copy stream). Source: User defined
    $sff_strict = 2 # `-strict <int>` ED.VA… how strictly to follow the standards (from INT_MIN to INT_MAX) (default 0). Source: User defined
    $sff_ac = 2 # `-ac <int>` channels set number of audio channels. Source: User defined
    $sff_ar = 44100 # `-ar <int>` rate set audio sampling rate (in Hz). Source: User defined
    # `-s <str>` size set frame size (WxH or abbreviation). Source: $_.T_height 
    $sff_map = 0 # `-map <int>` -map [-]input_file_id[:stream_specifier][,sync_file_id[:stream_s set input stream mapping. Source: User defined
    $sff_threads = 1 # `-threads <int>` (from 0 to INT_MAX) (default 1). Source: User Defined
    # `-v <string>` set logging level. Source: Built into command
    # `-stats` print progress report during encoding
    # `<outputpath>` output path of file being created. Source: $outputpath